### PR TITLE
TRUNK-5342: Migrate existing diagnosis data captured via the emrapi module

### PR DESCRIPTION
**Pull request by esirK**
## Description of what I changed

Created a Class to migrate existing diagnosis from the old obs table

## Issue I worked on

see https://issues.openmrs.org/browse/TRUNK-5342

## Checklist: I completed these to help reviewers :)

- [x] My pull request only contains **ONE single commit**
(the number above, next to the 'Commits' tab is 1).

  No? -> [read here](https://wiki.openmrs.org/display/docs/Pull+Request+Tips) on how to squash multiple commits into one

- [x] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [ ] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [x] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute above command

- [x] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [x] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

### CODE REVIEW

**Comment by dkayiwa** [comment]
@esirK i have just merged two pull requests in the emrapi and coreapps modules. Pull these changes and compile both modules. Then install them, test and give me feedback on if all is well.

File: api-2.2/src/main/java/org/openmrs/util/MigrateDiagnosis.java [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+	private org.openmrs.Diagnosis newDiagnosis;
+
+	public void migrate() {
+		List<Diagnosis> allOldDiagnosis;
+
+		allOldDiagnosis = oldDiagnosisService.getDiagnoses();
+
+		for (Diagnosis oldDiagnosis:allOldDiagnosis) {
+			newDiagnosis.setDiagnosis(oldDiagnosis.getDiagnosis());
```
Comment (Arthur236) [inline api-2.2/src/main/java/org/openmrs/util/MigrateDiagnosis.java]: Are the two diagnosis objects of the same type?

**Comment by CeciliaCaroline** [comment]
Add some comments to your work. I can not tell what the code is supposed to be doing.

File: api-2.2/src/main/java/org/openmrs/util/MigrateDiagnosis.java (hunk 1 above)
Comment (esirK) [reply api-2.2/src/main/java/org/openmrs/util/MigrateDiagnosis.java]: @Arthur236 Nope. They are not of the same type. oldDiagnosis  is from the emrapi which I want to migrate from while the newDiagnosis is the new diagnosis in the openmrs core.

**Comment by mogoodrich** [comment]
Is this complete or just a start?  This is generally the right path to take, but seems incomplete, as we want to copy all the properties of the old diagnosis (I haven't reviewed recently but I think some are missing) and then void the old diagnosis afterwards.

**Comment by esirK** [comment]
@mogoodrich yeah this is just the start I will be adding more properties

**Comment by mogoodrich** [comment]
Great @esirK ... that's what I figured, I just wasn't 100% sure.

File: api-2.2/src/main/java/org/openmrs/util/MigrateDiagnosis.java [hunk 2]
```diff
@@ -0,0 +1,94 @@
+/**
+ * The contents of this file are subject to the OpenMRS Public License
+ * Version 1.0 (the "License"); you may not use this file except in
```
Comment (dkayiwa) [inline api-2.2/src/main/java/org/openmrs/util/MigrateDiagnosis.java]: This is an older license.

File: api-2.2/src/main/java/org/openmrs/util/MigrateDiagnosis.java [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+	}
+
+	/**
+	 * Gets all patients with at-least a single diagnosis
+	 * @return all patients with at-least a diagnosis
+	 */
+	private List<Patient> getAllPatients(){
+		List<Patient> allPatients;
+
+		allPatients = Context.getPatientService().getPatients("select * from patient where patient_id in (select person_id from obs)");
```
Comment (dkayiwa) [inline api-2.2/src/main/java/org/openmrs/util/MigrateDiagnosis.java]: Are you sure this works?

Comment (esirK) [reply api-2.2/src/main/java/org/openmrs/util/MigrateDiagnosis.java]: @dkayiwa are you referring to the query or the whole method? If its the query, I tested it on my local database and its working

File: api-2.2/src/main/java/org/openmrs/util/MigrateDiagnosis.java (hunk 2 above)
Comment (esirK) [reply api-2.2/src/main/java/org/openmrs/util/MigrateDiagnosis.java]: Noted. Does this file require a license in the first place or can I just remove it??

Comment (dkayiwa) [reply api-2.2/src/main/java/org/openmrs/util/MigrateDiagnosis.java]: You just need to use the latest license

File: api-2.2/src/main/java/org/openmrs/util/MigrateDiagnosis.java (hunk 3 above)
Comment (mogoodrich) [reply api-2.2/src/main/java/org/openmrs/util/MigrateDiagnosis.java]: I will review in detail, but, to @dkayiwa point, there should be some unit tests around this.

File: api-2.2/src/main/java/org/openmrs/util/MigrateDiagnosis.java [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+	}
+
+	/**
+	 * Gets all patients_id s for all patients with at-least a diagnosis
+	 * @return list of patients_id s
+	 */
+	public List<Integer> getAllPatientsWithDiagnosis() {
+		List<Integer> allPatients = new ArrayList<Integer>();
+
+		Query query = sessionFactory.getCurrentSession().createQuery(
```
Comment (dkayiwa) [inline api-2.2/src/main/java/org/openmrs/util/MigrateDiagnosis.java]: You could put this in https://github.com/openmrs/openmrs-module-emrapi/blob/master/api/src/main/java/org/openmrs/module/emrapi/db/EmrVisitDAO.java
And https://github.com/openmrs/openmrs-module-emrapi/blob/master/api/src/main/java/org/openmrs/module/emrapi/db/EmrVisitDAOImpl.java

Comment (dkayiwa) [reply api-2.2/src/main/java/org/openmrs/util/MigrateDiagnosis.java]: The data access layer changes are accessed by the service layer via https://github.com/openmrs/openmrs-module-emrapi/blob/master/api/src/main/java/org/openmrs/module/emrapi/visit/EmrVisitService.java
And https://github.com/openmrs/openmrs-module-emrapi/blob/master/api/src/main/java/org/openmrs/module/emrapi/visit/EmrVisitServiceImpl.java

File: api-2.2/src/main/java/org/openmrs/util/MigrateDiagnosis.java [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+	/**
+	 * Creates new Diagnosis using the Diagnosis model from the openmrs-core and saves it using the new DiagnosisService
+	 *
+	 */
+	public void migrate() {
+		List<Diagnosis> allEmrapiDiagnosis = new ArrayList<Diagnosis>();
+
+		List<Integer> patientsIds;
+
+		patientsIds = emrVisitService.getAllPatientsWithDiagnosis();
```
Comment (dkayiwa) [inline api-2.2/src/main/java/org/openmrs/util/MigrateDiagnosis.java]: Isn't emrVisitService null at this time?

Comment (esirK) [reply api-2.2/src/main/java/org/openmrs/util/MigrateDiagnosis.java]: Sure it is null. Initializing it with `EmrVisitServiceImpl` will solve this right??

Comment (dkayiwa) [reply api-2.2/src/main/java/org/openmrs/util/MigrateDiagnosis.java]: You can use Context.getService

File: api/src/main/resources/hql/patients_diagnoses.hlq [hunk 6]
```diff
@@ -0,0 +1,7 @@
+select
```
Comment (paradisekelechi) [inline api/src/main/resources/hql/patients_diagnoses.hlq]: Is this supposed to be `patients_diagnoses.hlq`? @esirK

Comment (esirK) [reply api/src/main/resources/hql/patients_diagnoses.hlq]: Yes it is @paradisekelechi

File: api-2.2/src/main/java/org/openmrs/util/MigrateDiagnosis.java [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+			newDiagnosisService.save(diagnosis);
+		}
+	}
+
+	/**
+	 * Converts a list of emrapi diagnosis objects to a list of core diagnosis objects
+	 * @param emrapiDiagnoses list of emrapi diagnosis
+	 * @return a list of core diagnosis objects.
+	 */
+	private List<org.openmrs.Diagnosis> convert(List<Diagnosis> emrapiDiagnoses) {
```
Comment (paradisekelechi) [inline api-2.2/src/main/java/org/openmrs/util/MigrateDiagnosis.java]: I'm quite fascinated with this class. Looks pretty cool to me. Good job @esirK

File: api/src/main/resources/hql/patients_diagnoses.hlq (hunk 6 above)
Comment (paradisekelechi) [reply api/src/main/resources/hql/patients_diagnoses.hlq]: Kindly verify the file extension @esirK

**Comment by paradisekelechi** [comment]
Your continuous integration is failing. You want to check why?

File: api/src/main/resources/hql/patients_diagnoses.hlq (hunk 6 above)
Comment (esirK) [reply api/src/main/resources/hql/patients_diagnoses.hlq]: Thanks for this. Lemme fix it 😀

File: api-2.2/src/main/java/org/openmrs/util/MigrateDiagnosis.java [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+
+		DiagnosisService oldDiagnosisService = Context.getService(DiagnosisService.class);
+
+		org.openmrs.api.DiagnosisService newDiagnosisService = Context.getService(org.openmrs.api.DiagnosisService.class);
+
+		List<Diagnosis> allEmrapiDiagnosis = new ArrayList<Diagnosis>();
+
+		List<Integer> patientsIds;
+
+		patientsIds = emrVisitService.getAllPatientsWithDiagnosis();
```
[the remark this answers is not in the document]
Comment (esirK) [reply api-2.2/src/main/java/org/openmrs/util/MigrateDiagnosis.java]: Noted.

File: api-2.2/src/main/java/org/openmrs/util/MigrateDiagnosis.java [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+
+		List<Diagnosis> allEmrapiDiagnosis = new ArrayList<Diagnosis>();
+
+		List<Integer> patientsIds = emrVisitService.getAllPatientsWithDiagnosis();
+
+		for (int id: patientsIds) {
+			Patient patient= Context.getPatientService().getPatient(id);
+
+			//add all diagnoses of Patient patient to list of all emrapiDiagnosis
+			allEmrapiDiagnosis.addAll(oldDiagnosisService.getDiagnoses(patient, null));
```
Comment (mogoodrich) [inline api-2.2/src/main/java/org/openmrs/util/MigrateDiagnosis.java]: @dkayiwa  @djazayeri  do we care about migrating voided diagnoses?  If so we'll need a new version of this method that includes voided.  (I'm fine with keeping this as-is and not migrating voided diagnoses)

File: api-2.2/src/main/java/org/openmrs/util/MigrateDiagnosis.java [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+		for(Diagnosis emrapiDiagnosis: emrapiDiagnoses) {
+
+			Obs obs = emrapiDiagnosis.getExistingObs();
+
+			coreDiagnosis.setEncounter(obs.getEncounter());
+			coreDiagnosis.setPatient((Patient)obs.getPerson());
+			coreDiagnosis.setDiagnosis(new CodedOrFreeText(emrapiDiagnosis.getDiagnosis().getCodedAnswer(),
+					emrapiDiagnosis.getDiagnosis().getSpecificCodedAnswer(), emrapiDiagnosis.getDiagnosis().getNonCodedAnswer()));
+			coreDiagnosis.setCertainty(emrapiDiagnosis.getCertainty() == Diagnosis.Certainty.CONFIRMED ? ConditionVerificationStatus.CONFIRMED: ConditionVerificationStatus.PROVISIONAL);
+			coreDiagnosis.setUuid(obs.getUuid());
```
Comment (mogoodrich) [inline api-2.2/src/main/java/org/openmrs/util/MigrateDiagnosis.java]: This is the one thing we don't want to copy over... we should create a new, random uuid in this case. (Which the constructor will just do automatically I believe, so you can just remove this line).

File: api/src/main/resources/hql/patients_diagnoses.hql [hunk 11]
```diff
@@ -0,0 +1,7 @@
+select
+  distinct patient_id
+from
+  Patient p
+where
+  p.voided = 0
+  and p.patient_id in(select distinct person_id from obs)
```
Comment (mogoodrich) [inline api/src/main/resources/hql/patients_diagnoses.hql]: Note that it appears that the hql query will return all patients that have any obs, not just those with diagnoses, you would want to limit this to patients that have at least one obs where obs.concept_id = : diagnosisSetConceptId.  (You will have to retrieve the diagnosis set concept using the DiagnosisMetadata service and inject the ID into the query similar to the way other values are injected in EmrVisitDAOImpl.

**Comment by esirK** [comment]
@mogoodrich Thank you very much for the comments. I am working on them now. I would also appreciate some pointers on how to start writing the unit tests since I have been researching on the same but no success yet.

File: api-2.2/src/main/java/org/openmrs/util/MigrateDiagnosis.java (hunk 10 above)
Comment (esirK) [reply api-2.2/src/main/java/org/openmrs/util/MigrateDiagnosis.java]: Noted

**Comment by mogoodrich** [comment]
@esirK  you can take a look at the DiagnosisServiceComponentTest in the emr-api module as a start.  You should be able to use the "buildDiagnosis" method in this class to use a diagnosis in the old format and persist it to the database.  Then you would execute your merge code, and afterwards confirm that 1) the old diagnosis has been voided, and 2) the new diagnosis has been created.

**Comment by mogoodrich** [comment]
@esirK  my last comment reminded me... does your migration code void the old diagnosis obs?

**Comment by esirK** [comment]
@mogoodrich thanks for the pointers on where to start. For the voiding, I have just updated the code with the same.

File: pom.xml [hunk 12]
```diff
@@ -1,65 +1,65 @@
 <project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
-    <modelVersion>4.0.0</modelVersion>
-    <groupId>org.openmrs.module</groupId>
-    <artifactId>emrapi</artifactId>
-    <version>1.25.0-SNAPSHOT</version>
-    <packaging>pom</packaging>
```
Comment (dkayiwa) [inline pom.xml]: Can you revert the formatting changes to this file? We only need to see what actually changed.

File: api-2.2/src/test/java/org/openmrs/util/MigrateDiagnosisTest.java [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+	private DiagnosisMetadata diagnosisMetadata;
+
+	@Before
+	public void setUp() throws Exception {
+		executeDataSet(DIAGNOSIS_DATASET);
+		diagnosisMetadata = ContextSensitiveMetadataTestUtils.setupDiagnosisMetadata(conceptService, emrApiProperties);
+	}
+
+	@Test
+	public void getAllPatientsWithDiagnosisShouldReturnListOfDiagnosis() {
```
Comment (paradisekelechi) [inline api-2.2/src/test/java/org/openmrs/util/MigrateDiagnosisTest.java]: @esirK 
This method name does not represent what it does.

**Comment by samuelmale** [comment]
@esirK  Are you still working on this?

**Comment by esirK** [comment]
@samuelmale currently not.

**Comment by samuelmale** [comment]
@esirK  Then I'm closing it. When you feel like you want to do more work, you can go on.

**Comment by dkayiwa** [comment]
@esirK did you get a chance to test this by running a real application and attempt to do a migration?

**Comment by dkayiwa** [comment]
Closing this because @esirK is no longer working on it.

**Pull request closed without merging**